lib/unification.h

```diff
@@ -1,52 +1,123 @@
+#include "fol/Literal.h" // negatable predicate
+#include "fol/Predicate.h" //predicate class, used as "label" of literal
+#include "Visitor.h" // for type checking
 
-using Substitution = std::unordered_map<std::string, fol::Constant>;
-
-// Forward declaration
-template <class T>
-Substitution unify_var(fol::Variable& var, T x, Substitution& theta);
-
-template <class T>
-auto unify(fol::Variable& x, T& y, std::optional<Substitution>& theta) {
-    return theta ? unify_var(x, y, theta) : std::nullopt;
-}
-
-template <class T, class U> auto unify(T& x, U& y) {
-    auto subst = std::optional<Substitution>();
-    return unify(x, y, subst);
-}
+using namespace std;
+using namespace fol;
 
-template <class T> auto unify(T& x, T& y) {
-    auto subst = std::optional<Substitution>();
-    return unify(x, y, subst);
-}
+// setting up unordered_map for our substitution list
+typedef unordered_map<string, Term> sub_list_type;
+using sub_list = variant<sub_list_type, string>;
 
-auto unify(fol::Constant& x, fol::Constant& y, std::optional<Substitution>& theta) {
-    return theta ? theta : std::nullopt;
+// now for the substition formula for replacing the variable in an atom/predicate
+void substitute(Literal<Term> &x, Literal<Term> &y, sub_list &z, int ix) {
+    if (boost::apply_visitor(type_visitor(), (y.args).at(ix)) == "Variable") {
+        // write the substitution to z
+        get<sub_list_type>(z)[get<Variable>((x.args).at(ix)).name] = get<Variable>((y.args).at(ix));
+        // insert the new subbed element and then erase the old one
+        (x.args).insert(x.args.begin() + ix, get<Variable>((y.args).at(ix)));
+        ix++;
+        (x.args).erase(x.args.begin() + ix);
+    }
+    else if (boost::apply_visitor(type_visitor(), (y.args).at(ix)) == "Constant") {
+        // write substitution to z
+        get<sub_list_type>(z)[get<Variable>((x.args).at(ix)).name] = get<Constant>((y.args).at(ix));
+        // insert substitution and then erase the old entry
+        (x.args).insert(x.args.begin() + ix, get<Constant>((y.args).at(ix)));
+        ix++;
+        (x.args).erase(x.args.begin() + ix);
+    }
+    else if (boost::apply_visitor(type_visitor(), (y.args).at(ix)) == "Predicate") {
+        // write substitution to z
+        get<sub_list_type>(z)[get<Variable>((x.args).at(ix)).name] = get<Function>((y.args).at(ix));
+        // insert substitution and then erase the old entry
+        (x.args).insert(x.args.begin() + ix, get<Function>((y.args).at(ix)));
+        ix++;
+        (x.args).erase(x.args.begin() + ix);
+    }
 }
 
-template <class T>
-std::optional<Substitution>
-unify_var(fol::Variable& var, T x, std::optional<Substitution>& theta) {
-    if (in(var.name, theta.value())) {
-        auto val = theta.value()[var.name];
-        return unify(val, x, theta);
+// now we apply the unification algorithm on the predicate pairs
+sub_list unification(Literal<Term> x, Literal<Term> y, sub_list z) {
```

Comment:
Can the `unification` function be renamed to `unify`? I usually like to give functions names that sound like actions (i.e. simple present tense)...

Reply:
No problem! If I push a commit will it update the pull request? or do I have to close the request, push the commit, and then open another request?

Reply:
@Free-Quarks you can just push a commit and the PR will automatically get updated!

Reply:
should be updated now!

Reply:
Great, thanks! I'll merge after tests pass.